### PR TITLE
Handle free agent assignment and removal

### LINKED ISSUE #126: Remove free agents when they join a team
When a user that is signed up as free agent for a season joins a team that is signed up for that season, or when a team that the user is in signs up for that season, the user should be removed from that seasons free agent list

### PR DESCRIPTION
**Pull request by Astraeus-**
Fixes #126 and #140.

Users will be removed from the Free Agent list if they join a team signed up to a season with open registration or one of their teams signs up for a season.

Also allows users signed up as a Free Agent, but not participating with a team to remove their sign up from a season with open registration.

### CODE REVIEW

File: plugins/rikki/heroeslounge/components/SlothAccount.php [hunk 1]
```diff
@@ -279,26 +279,27 @@ public function onRegister()
...
-                        $season->free_agents()->attach($this->sloth->id);
-                        Discord\RoleManagement::UpdateUserRole("PUT", $this->sloth->discord_id, "FreeAgent");
-                        Flash::success('You will participate in '.$season->title.'!');
-                    } elseif (isset($val) == false) {
-                        $season->free_agents()->detach($this->sloth->id);
-                        Discord\RoleManagement::UpdateUserRole("DELETE", $this->sloth->discord_id, "FreeAgent");
-                        Flash::error('You will not paricipate in '.$season->title.' - Sad to see you go!');
-                    }
+        foreach ($this->seasons as $season) {
+            if ($season->reg_open == 1) {
```
Comment (Fabian-Sommer) [inline plugins/rikki/heroeslounge/components/SlothAccount.php]: You can leave away the == 1

File: plugins/rikki/heroeslounge/components/ViewApps.php [hunk 2]
```diff
@@ -118,6 +120,15 @@ public function onAccept()
                     //a sloth should not participate in two different teams in the same season or tournament
                     if ($this->sloth->mayJoinTeam($team)) {
                         $this->sloth->teams()->add($team);
+
+                        $seasons = Season::all();
+                        foreach ($seasons as $season) {
+                            if ($season->reg_open == 1 && $season->teams->contains($team) && $season->free_agents->contains($this->sloth)) {
```
Comment (Fabian-Sommer) [inline plugins/rikki/heroeslounge/components/ViewApps.php]: == 1

File: plugins/rikki/heroeslounge/components/SlothAccount.php [hunk 3]
```diff
@@ -279,26 +279,27 @@ public function onRegister()
...
+                    Flash::error('You will not participate in '.$season->title.' - Sad to see you go!');
                 }
             }
-        } catch (Exception $e) {
-            Flash::error($e->getMessage());
-        } finally {
-            return Redirect::refresh();
         }
+
+        return Redirect::refresh();
```
Comment (Fabian-Sommer) [inline plugins/rikki/heroeslounge/components/SlothAccount.php]: General comment: We can just get rid of the entire "Sign up on your profile" concept

Comment (Astraeus-) [reply plugins/rikki/heroeslounge/components/SlothAccount.php]: So would we still want users to be able to remove their participation from a season on their profile? Or would we want this to also be removed and added somewhere else?

Comment (Fabian-Sommer) [reply plugins/rikki/heroeslounge/components/SlothAccount.php]: Ideally, all of the signing up / remove participation should be on the season / playoff page

Comment (Astraeus-) [reply plugins/rikki/heroeslounge/components/SlothAccount.php]: Updated the free agent participation changes to be on season pages. I did keep a listing of participation on the profile page for legacy reasons.

I didn't change anything for playoffs, as we don't allow free agent sign ups there anyway.

File: plugins/rikki/heroeslounge/components/ViewApps.php [hunk 4]
```diff
@@ -118,6 +120,15 @@ public function onAccept()
                     //a sloth should not participate in two different teams in the same season or tournament
                     if ($this->sloth->mayJoinTeam($team)) {
                         $this->sloth->teams()->add($team);
+
+                        $seasons = Season::all();
+                        foreach ($seasons as $season) {
+                            if ($season->reg_open && $season->teams->contains($team) && $season->free_agents->contains($this->sloth)) {
+                                $season->free_agents()->remove($this->sloth);
```
Comment (Fabian-Sommer) [inline plugins/rikki/heroeslounge/components/ViewApps.php]: Why did you change detach to remove?

File: plugins/rikki/heroeslounge/components/slothaccount/update.htm [hunk 5]
```diff
@@ -286,29 +285,35 @@ <h3>Server Preference</h3>
...
+            <div>
+                <label>
+                    <input type="checkbox" name="part_seas_{{season.id}}" {% if checked %} checked {% endif %} disabled>
+                    {{season.title}}
+                </label>
+            </div>
+            {% set checked = '0' %}
+        {% endfor %}
 
         <!-- END PARTICIPATION -->
```
Comment (Fabian-Sommer) [inline plugins/rikki/heroeslounge/components/slothaccount/update.htm]: Please remove this from this page

File: plugins/rikki/loungeviews/components/ParticipationOverview.php [hunk 6]
```diff
@@ -99,9 +105,25 @@ public function onSlothSignup()
...
         }
     }
+
+    public function onSlothRemoveSignUp()
+    {
+        $this->user = Auth::getUser();
+        $this->season = Seasons::find($_POST['season_id']);
+        if ($this->user != null) {
+            $sloth = $this->user->sloth;
+            if ($sloth != null && $sloth->region_id == $this->season->region_id && $this->season->free_agents->contains($sloth)) {
```
Comment (Fabian-Sommer) [inline plugins/rikki/loungeviews/components/ParticipationOverview.php]: All of these checks should be unnecessary.

File: plugins/rikki/loungeviews/components/participationoverview/default.htm [hunk 7]
```diff
@@ -30,10 +30,20 @@ <h1>{{__SELF__.season.title}} Participation</h1>
                 class="btn btn-sm btn-secondary center"
                 data-request-data="season_id: {{__SELF__.season.id}}"
                 data-request="{{ __SELF__ }}::onSlothSignup">
-            Sign up as a free agent!
+            Sign up as a free agent
         </button>
         <br><br>
     </div>
+    {% elseif __SELF__.user and __SELF__.signedUp and __SELF__.user.sloth.region_id == __SELF__.season.region_id %}
```
Comment (Fabian-Sommer) [inline plugins/rikki/loungeviews/components/participationoverview/default.htm]: Why do you check for the right region id here?

File: plugins/rikki/loungeviews/components/participationoverview/default.htm [hunk 8]
```diff
@@ -30,10 +30,20 @@ <h1>{{__SELF__.season.title}} Participation</h1>
...
+    <div>
+        No longer want to particpate as a free agent?<br>
```
Comment (Fabian-Sommer) [inline plugins/rikki/loungeviews/components/participationoverview/default.htm]: Is there something like 'You are signed up as a free agent' on this page? If not, please add it before this.

File: plugins/rikki/heroeslounge/components/ViewApps.php (hunk 4 above)
Comment (Astraeus-) [reply plugins/rikki/heroeslounge/components/ViewApps.php]: Since we use the adding equivalent when a sloth signs up for a season. I can use detach if you prefer though.

Comment (Fabian-Sommer) [reply plugins/rikki/heroeslounge/components/ViewApps.php]: Remove is fine if it works

Comment (Astraeus-) [reply plugins/rikki/heroeslounge/components/ViewApps.php]: I tested it locally and it worked as expected

**Comment by Astraeus-** [comment]
Noticed in final testing that checking signup was general and not specific to free agent, so made sure that the user is actually a free agent to allow them to sign up or no longer participate
